### PR TITLE
[IIP-13] introduce read contract staking buckets (2/3 split by 3853) 

### PR DESCRIPTION
**Pull request by envestcc**
# Description
For #3853 is too large to review, it has been splited into three small pr. This is the second one.
It's based on #3861 

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/contractstaking/indexer.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	// Indexer is the contract staking indexer
+	// Main functions:
+	// 		1. handle contract staking contract events when new block comes to generate index data
+	// 		2. provide query interface for contract staking index data
+	// Generate index data flow:
+	// 		block comes -> new dirty cache -> handle contract events -> update dirty cache -> merge dirty to clean cache
+	// Main Object:
+	// 		kvstore: persistent storage, used to initialize index cache at startup
+	// 		cache: in-memory index for clean data, used to query index data
+	//      dirty: the cache to update during event processing, will be merged to clean cache after all events are processed. If errors occur during event processing, dirty cache will be discarded.
```
Comment (CoderZhi) [inline action/protocol/staking/contractstaking/indexer.go:35]: comment above variables

File: action/protocol/staking/contractstaking/indexer.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+	maxBlockNumber uint64 = math.MaxUint64
+)
+
+type (
+	// Indexer is the contract staking indexer
+	// Main functions:
+	// 		1. handle contract staking contract events when new block comes to generate index data
+	// 		2. provide query interface for contract staking index data
+	// Generate index data flow:
```
Comment (CoderZhi) [inline action/protocol/staking/contractstaking/indexer.go:30]: flow explanation should not be here

File: action/protocol/staking/contractstaking/cache.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	if !ok {
+		return votes
+	}
+	for id, existed := range m {
+		if !existed {
+			continue
+		}
+		bi, ok := s.idBucketMap[id]
+		if !ok {
+			continue
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:57]: don't consider this an error? bucket exist in candidate's map, but not exist in total map

File: action/protocol/staking/contractstaking/bucket.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		StakeStartBlockHeight:     bi.CreatedAt,
+		UnstakeStartBlockHeight:   bi.UnstakedAt,
+		AutoStake:                 bi.UnlockedAt == maxBlockNumber,
+		Candidate:                 bi.Delegate,
+		Owner:                     bi.Owner,
+		ContractAddress:           StakingContractAddress,
+	}
+	if bi.UnlockedAt != maxBlockNumber {
+		vb.StakeStartBlockHeight = bi.UnlockedAt
+	}
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/bucket.go:43]: what is the logic here?

File: action/protocol/staking/contractstaking/cache.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+)
+
+func newContractStakingCache() *contractStakingCache {
+	cache := &contractStakingCache{
+		idBucketMap:           make(map[uint64]*bucketInfo),
+		idBucketTypeMap:       make(map[uint64]*BucketType),
+		propertyBucketTypeMap: make(map[int64]map[uint64]uint64),
+		candidateBucketMap:    make(map[string]map[uint64]bool),
+	}
+	return cache
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:38]: nit:
```
cache := contractStakingCache{
return &cache
```

File: action/protocol/staking/contractstaking/cache.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	for id, existed := range m {
+		if !existed {
+			continue
+		}
+		bi, ok := s.idBucketMap[id]
+		if !ok {
+			continue
+		}
+		if bi.UnstakedAt != maxBlockNumber {
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:59]: what is the logic here

File: action/protocol/staking/contractstaking/cache.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+func (s *contractStakingCache) GetTotalBucketCount() uint64 {
+	return s.totalBucketCount
+}
+
+func (s *contractStakingCache) GetActiveBucketTypes() map[uint64]*BucketType {
+	m := make(map[uint64]*BucketType)
+	for k, v := range s.idBucketTypeMap {
+		if v.ActivatedAt != maxBlockNumber {
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:117]: what is the logic here

File: action/protocol/staking/contractstaking/cache.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+)
+
+type (
+	contractStakingCache struct {
+		idBucketMap           map[uint64]*bucketInfo      // map[token]bucketInfo
+		candidateBucketMap    map[string]map[uint64]bool  // map[candidate]bucket
+		idBucketTypeMap       map[uint64]*BucketType      // map[token]BucketType
+		propertyBucketTypeMap map[int64]map[uint64]uint64 // map[amount][duration]index
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:20]: what is `index`?

File: action/protocol/staking/contractstaking/cache.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+)
+
+type (
+	contractStakingCache struct {
+		idBucketMap           map[uint64]*bucketInfo      // map[token]bucketInfo
+		candidateBucketMap    map[string]map[uint64]bool  // map[candidate]bucket
+		idBucketTypeMap       map[uint64]*BucketType      // map[token]BucketType
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:19]: just `bucketTypeMap`? the `id` here is NOT the same as `id` in L17, right? If yes, then the naming could be a little confusing

File: action/protocol/staking/contractstaking/cache.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	bt := s.mustGetBucketType(bi.TypeIndex)
+	return assembleBucket(id, bi, bt)
+}
+
+func (s *contractStakingCache) getTotalBucketTypeCount() uint64 {
+	return uint64(len(s.idBucketTypeMap))
+}
+
+func (s *contractStakingCache) getAllBucketInfo() map[uint64]*bucketInfo {
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:172]: cannot directly return `s.idBucketMap`?
if `s.idBucketMap` will be accessed/modified at the same time, then this copy won't help
if `s.idBucketMap` won't be accessed/modified, then might as well just return `s.idBucketMap`
either way, copying the map does not seem to help

File: action/protocol/staking/contractstaking/cache.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	m := make(map[uint64]*bucketInfo)
+	for k, v := range s.idBucketMap {
+		m[k] = v
+	}
+	return m
+}
+
+func (s *contractStakingCache) getBucketInfoByCandidate(candidate address.Address) map[uint64]*bucketInfo {
+	m := make(map[uint64]*bucketInfo)
```
Comment (dustinxie) [inline action/protocol/staking/contractstaking/cache.go:181]: same question as above

**Review by dustinxie — APPROVED** [review]

File: action/protocol/staking/contractstaking/cache.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+	cache := &contractStakingCache{
+		idBucketMap:           make(map[uint64]*bucketInfo),
+		idBucketTypeMap:       make(map[uint64]*BucketType),
+		propertyBucketTypeMap: make(map[int64]map[uint64]uint64),
+		candidateBucketMap:    make(map[string]map[uint64]bool),
+	}
+	return cache
+}
+
+func (s *contractStakingCache) GetHeight() uint64 {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:41]: fixed

File: action/protocol/staking/contractstaking/cache.go [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+	return s.height
+}
+
+func (s *contractStakingCache) GetCandidateVotes(candidate address.Address) *big.Int {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:45]: fixed

File: action/protocol/staking/contractstaking/cache.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:57]: use `mustGetBucketInfo` instead

File: action/protocol/staking/contractstaking/cache.go [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+		if bi.UnstakedAt != maxBlockNumber {
+			continue
+		}
+		bt := s.mustGetBucketType(bi.TypeIndex)
+		votes.Add(votes, bt.Amount)
+	}
+	return votes
+}
+
+func (s *contractStakingCache) GetBuckets() ([]*Bucket, error) {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:68]: fixed

File: action/protocol/staking/contractstaking/cache.go [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+		vb, err := assembleBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+func (s *contractStakingCache) GetBucket(id uint64) (*Bucket, error) {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:81]: fixed

File: action/protocol/staking/contractstaking/cache.go [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+	return s.getBucket(id)
+}
+
+func (s *contractStakingCache) GetBucketsByCandidate(candidate address.Address) ([]*Bucket, error) {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:85]: fixed

File: action/protocol/staking/contractstaking/cache.go [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+		vb, err := s.getBucket(id)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+func (s *contractStakingCache) GetBucketsByIndices(indices []uint64) ([]*Bucket, error) {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:98]: fixed

File: action/protocol/staking/contractstaking/cache.go [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+func (s *contractStakingCache) GetTotalBucketCount() uint64 {
+	return s.totalBucketCount
+}
+
+func (s *contractStakingCache) GetActiveBucketTypes() map[uint64]*BucketType {
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:114]: fixed

File: action/protocol/staking/contractstaking/indexer.go [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/iotexproject/iotex-address/address"
+
+	"github.com/iotexproject/iotex-core/db"
+)
+
+const (
+	// StakingContractAddress  is the address of system staking contract
+	// TODO (iip-13): replace with the real system staking contract address
+	StakingContractAddress = "io19ys8f4uhwms6lq6ulexr5fwht9gsjes8mvuugd"
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/indexer.go:20]: fixed

File: action/protocol/staking/contractstaking/indexer.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/indexer.go:30]: removed and should be added in #3863 at `PutBlock`

File: action/protocol/staking/contractstaking/indexer.go (hunk 1 above)
Comment (envestcc) [reply action/protocol/staking/contractstaking/indexer.go:35]: fixed

**Comment by millken** [comment]
add integration testing,improve test coverage

**Comment by envestcc** [comment]
> add integration testing,improve test coverage

tests are added in next PR #3863

**Review by millken — APPROVED** [review]

File: action/protocol/staking/contractstaking/indexer.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/indexer.go]: will be implemented in the integrate PR

File: action/protocol/staking/contractstaking/cache.go [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+	return id, ok
+}
+
+func (s *contractStakingCache) getBucketType(id uint64) (*BucketType, bool) {
+	bt, ok := s.bucketTypeMap[id]
+	return bt, ok
+}
+
+func (s *contractStakingCache) mustGetBucketType(id uint64) *BucketType {
+	bt, ok := s.bucketTypeMap[id]
```
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/contractstaking/cache.go:139]: sure

**Pull request merged**